lib/evmone/eof.cpp

```diff
@@ -90,37 +151,261 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+            if (jumpdest < 0 || static_cast<size_t>(jumpdest) >= code_size)
+                return false;
+            rjumpdests.push_back(static_cast<size_t>(jumpdest));
+        }
+        else if (op == OP_RJUMPV)
+        {
+            constexpr auto REL_OFFSET_SIZE = sizeof(int16_t);
+
+            const auto count = code[i + 1];
+
+            const auto post_offset = 1 + 1 /* count */ + count * REL_OFFSET_SIZE /* tbl */;
+
+            for (size_t k = 0; k < count * REL_OFFSET_SIZE; k += REL_OFFSET_SIZE)
+            {
+                const auto rel_offset = read_int16_be(&code[i + 1 + 1 + static_cast<uint16_t>(k)]);
+                const auto jumpdest = static_cast<int>(i + post_offset) + rel_offset;
+                if (jumpdest < 0 || static_cast<size_t>(jumpdest) >= code_size)
+                    return false;
+                rjumpdests.push_back(static_cast<size_t>(jumpdest));
+            }
+
+            const auto rjumpv_imm_size = size_t{1} + count * REL_OFFSET_SIZE;
+            std::fill_n(
+                immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, rjumpv_imm_size, true);
+            i += rjumpv_imm_size;
+            continue;
+        }
+
+        const auto imm_size = instr::traits[op].immediate_size;
+        std::fill_n(immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, imm_size, true);
+        i += imm_size;
+    }
+
+    // Check relative jump destinations against immediate locations.
+    for (const auto rjumpdest : rjumpdests)
+        if (immediate_map[rjumpdest])
+            return false;
+
+    return true;
+}
+
+std::pair<EOFValidationError, int32_t> validate_max_stack_height(
+    bytes_view code, size_t func_index, const std::vector<EOFCodeType>& funcs_in_outs)
+{
+    assert(!code.empty());
+
+    // Changing this type to smaller integer (uint16_t) have in mind that it can overflow.
+    std::vector<int32_t> stack_heights = std::vector<int32_t>(code.size(), -1);
+    std::stack<size_t> worklist;
+
+    int32_t stack_height = 0;
+    stack_heights[0] = funcs_in_outs[func_index].inputs;
+    worklist.push(0);
+
+    size_t i = 0;
+    Opcode opcode = {};
+    std::vector<size_t> successors;
+    while (!worklist.empty())
+    {
+        i = worklist.top();
+        opcode = static_cast<Opcode>(code[i]);
+        worklist.pop();
+
+        auto stack_height_required = instr::traits[opcode].stack_height_required;
+        auto stack_height_change = instr::traits[opcode].stack_height_change;
+
+        if (opcode == OP_CALLF)
+        {
+            auto fid = read_uint16_be(&code[i + 1]);
+
+            if (fid >= funcs_in_outs.size())
+                return {EOFValidationError::invalid_code_section_index, 0};
+
+            stack_height_required = static_cast<int8_t>(funcs_in_outs[fid].inputs);
+            auto d = funcs_in_outs[fid].outputs - stack_height_required;
+            stack_height_change = static_cast<int8_t>(d);
```

Comment:
Hm, why not the following to signal where is truncation happening?
```suggestion
            stack_height_change = static_cast<int8_t>(funcs_in_outs[fid].outputs) - stack_height_required;
```

Reply:
Done

Reply:
```
stack_height_change =
                static_cast<int8_t>(funcs_in_outs[fid].outputs - stack_height_required);
```

The new code casts both, my point was only casting the input, i.e. parentheses close after `)`.

Reply:
Because of sanitizer build error:
```
/home/builder/project/lib/evmone/eof.cpp:340:17: error: narrowing conversion from 'int' to signed type 'int8_t' (aka 'signed char') is implementation-defined [bugprone-narrowing-conversions,cppcoreguidelines-narrowing-conversions,-warnings-as-errors]
                static_cast<int8_t>(funcs_in_outs[fid].outputs) - stack_height_required;
```
I changed this to `stack_height_change` type to int32_t

Reply:
I think the sanitizer is wrong, `outputs` is limited to the range of `int8` already, and `stack_height_required` is already that type. This should work imho.

Reply:
Unfortunately, C/C++ promotes smaller types to `int` for arithmetic operations. So the result of `-` is `int`.